src/backend/distributed/commands/non_main_db_distribute_object_ops.c

```diff
@@ -0,0 +1,450 @@
+	appendStringInfo(mainDBQuery,
+					 START_MANAGEMENT_TRANSACTION,
+					 GetCurrentFullTransactionId().value);
+	RunCitusMainDBQuery(mainDBQuery->data);
+
+	mainDBQuery = makeStringInfo();
+	appendStringInfo(mainDBQuery,
+					 EXECUTE_COMMAND_ON_REMOTE_NODES_AS_USER,
+					 quote_literal_cstr(queryString),
+					 quote_literal_cstr(CurrentUserName()));
+	RunCitusMainDBQuery(mainDBQuery->data);
+
+	if (ops->getUnmarkDistributedParams)
+	{
+		List *unmarkDistributedParamsList = ops->getUnmarkDistributedParams(parsetree);
+		UnmarkObjectsDistributedOnLocalMainDb(unmarkDistributedParamsList);
+	}
+
+	return false;
+}
+
+
+/*
+ * RunPostprocessNonMainDBCommand runs the necessary commands for a query, in main
+ * database after query is run on the local node with PrevProcessUtility.
+ */
+void
+RunPostprocessNonMainDBCommand(Node *parsetree)
+{
+	if (IsMainDB)
+	{
+		return;
+	}
+
+	const NonMainDbDistributeObjectOps *ops = GetNonMainDbDistributeObjectOps(parsetree);
+	if (!ops)
+	{
+		return;
+	}
+
+	if (ops->getMarkDistributedParams)
+	{
+		MarkDistributedGloballyParams *markDistributedParams =
+			ops->getMarkDistributedParams(parsetree);
+		MarkObjectDistributedGloballyOnMainDbs(markDistributedParams);
+	}
+}
+
+
+/*
+ * GetNonMainDbDistributeObjectOps returns the NonMainDbDistributeObjectOps for given
+ * command if it's node-wide object management command that's supported from non-main
+ * databases.
+ */
+const NonMainDbDistributeObjectOps *
+GetNonMainDbDistributeObjectOps(Node *parsetree)
+{
+	NodeTag tag = nodeTag(parsetree);
+	if (tag >= lengthof(operationArray))
+	{
+		return NULL;
+	}
+
+	const NonMainDbDistributeObjectOps *ops = operationArray[tag];
+
+	switch (nodeTag(parsetree))
+	{
+		case T_CreatedbStmt:
+		{
+			CreatedbStmt *stmt = castNode(CreatedbStmt, parsetree);
+
+			/*
+			 * We don't try to send the query to the main database if the CREATE
+			 * DATABASE command is for the main database itself, this is a very
+			 * rare case but it's exercised by our test suite.
+			 */
+			if (strcmp(stmt->dbname, MainDb) != 0)
+			{
+				return ops;
+			}
+
+			return NULL;
+		}
+
+		case T_DropdbStmt:
+		{
+			DropdbStmt *stmt = castNode(DropdbStmt, parsetree);
+
+			/*
+			 * We don't try to send the query to the main database if the DROP
+			 * DATABASE command is for the main database itself, this is a very
+			 * rare case but it's exercised by our test suite.
+			 */
+			if (strcmp(stmt->dbname, MainDb) != 0)
+			{
+				return ops;
+			}
+
+			return NULL;
+		}
+
+		case T_GrantStmt:
+		{
+			GrantStmt *stmt = castNode(GrantStmt, parsetree);
+
+			switch (stmt->objtype)
+			{
+				case OBJECT_DATABASE:
+				{
+					return ops;
+				}
+
+				default:
+					return NULL;
+			}
+		}
+
+		case T_SecLabelStmt:
+		{
+			SecLabelStmt *stmt = castNode(SecLabelStmt, parsetree);
+
+			switch (stmt->objtype)
+			{
+				case OBJECT_ROLE:
+				{
+					return ops;
+				}
+
+				default:
+					return NULL;
+			}
```

Comment:
This seems quite verbose for what was first these easy functions:
```c
/*
 * NonMainDbCheckSupportedObjectTypeForGrant implements checkSupportedObjectTypes
 * callback for GrantStmt.
 */
static bool
NonMainDbCheckSupportedObjectTypeForGrant(Node *node)
{
	GrantStmt *stmt = castNode(GrantStmt, node);
	return stmt->objtype == OBJECT_DATABASE;
}


/*
 * NonMainDbCheckSupportedObjectTypeForSecLabel implements checkSupportedObjectTypes
 * callback for SecLabel.
 */
static bool
NonMainDbCheckSupportedObjectTypeForSecLabel(Node *node)
{
	SecLabelStmt *stmt = castNode(SecLabelStmt, node);
	return stmt->objtype == OBJECT_ROLE;
}
```

If we keep these functions and make similar functions for the code in the `T_CreatedbStmt` and `T_DropdbStmt` blocks, then the `switch` statement can suddenly be completely removed when we add a `nonMainDBoperationCheck` callback to the ops struct that returns false for the ones that cases where nothing should happen.

Reply:
Similarly, I really liked the idea in your commit to reduce the line-of-code in the switch-case statement but I'm not much in favor of completely removing the switch-case and doing the checks in the call-site. For example, for "alter database" commands, we also need to skip processing the command from non-main dbs if it targets main db and I don't much want to clutter the call-site.

Reply:
To be clear, for this case I would actually like an abstraction in `NonMainDbDistributeObjectOps`. So `GetNonMainDbDistributeObjectOps` would then look something like this (name of callback to be bikeshedded):

```c
const NonMainDbDistributeObjectOps *
GetNonMainDbDistributeObjectOps(Node *parsetree)
{
	NodeTag tag = nodeTag(parsetree);
	if (tag >= lengthof(operationArray))
	{
		return NULL;
	}

	const NonMainDbDistributeObjectOps *ops = operationArray[tag];

	if (ops == NULL)
	{
		return NULL;
	}
	
	if (ops->shouldForwardToMainDb())
	{
		return ops;
	}

	return NULL;
}
```